src/hotspot/share/opto/mulnode.cpp

```diff
@@ -1389,7 +1394,21 @@ const Type* RShiftLNode::Value(PhaseGVN* phase) const {
     return tl;
   }
 
-  return TypeLong::LONG;                // Give up
+  if (!r1->is_con() || !r2->is_con()) {
+    // If the left input is non-negative the result must also be non-negative, regardless of what the right input is.
+    if (r1->_lo >= 0) {
+      return TypeLong::make(0, r1->_hi, MAX2(r1->_widen, r2->_widen));
+    }
+
+    // Conversely, if the left input is negative then the result must be negative.
+    if (r1->_hi <= -1) {
+      return TypeLong::make(r1->_lo, -1, MAX2(r1->_widen, r2->_widen));
+    }
+
+    return TypeLong::LONG;
+  }
+
+  return TypeLong::make((jlong)r1->get_con() >> (r2->get_con() & 63));
```

Comment:
The cast to `jlong` can be removed since `get_con()` already returns a `jlong`.
```suggestion
  return TypeLong::make(r1->get_con() >> (r2->get_con() & 63));
```

Reply:
Nice catch, thank you!